pycbc/events/stat.py

```diff
@@ -1758,6 +1757,155 @@ def coinc_lim_for_thresh(self, s, thresh, limifo,
+            The name of the ranking to use for the single-detector triggers.
+        files: list of strs, needed here
+            A list containing the filenames of hdf format files used to help
+            construct the coincident statistics. The files must have a 'stat'
+            attribute which is used to associate them with the appropriate
+            statistic class.
+        ifos: list of strs, not used here
+            The list of detector names
+        """
+        ExpFitFgBgNormStatistic.__init__(self, sngl_ranking, files=files,
+                                         ifos=ifos, **kwargs)
+        parsed_attrs = [f.split('-') for f in self.files.keys()]
+        self.data_name = [at[0] for at in parsed_attrs if
+                       (len(at) == 2 and at[1] == 'kde_file')]
+        if not len(self.data_name):
+            raise RuntimeError("None of the statistic files has the required "
+                               "attribute called {dname}-kde_file !")
+        self.kde_by_tid = {}
+        for dname in self.data_name:
+            self.kde_by_tid[dname] = self.assign_kdes(dname)
+            #print('self.kde_by_tid=', self.kde_by_tid[dname])
+        self.curr_tnum = None       
+ 
+    def assign_kdes(self, dname):
+        """
+        Extract fits from fit files
+
+        Parameters
+        -----------
+        dname: str
+            signal or template.
+
+        Returns
+        -------
+        rate_dict: dict
+            A dictionary containing kde values for signal and template files
+
+        """
+        kde_file = self.files[dname+'-kde_file']
+        template_id = kde_file['template_id'][:]
+        tid_sort = numpy.argsort(template_id)
+        kde_by_tid_dict = {}
+        kde_by_tid_dict[dname+'_data_kde'] = \
+            kde_file['data_kde'][:][tid_sort]
+        print('self.kde_by_tid_dict', kde_by_tid_dict)
+        return kde_by_tid_dict
+
+    def rank_stat_coinc(self, stats, shift, to_shift):
+        """
+        Calculate the normalized log rate density of signals via lookup
+
+        This calls back to the Parent class and then applies the ratio_kde
+        weighting factor.
+
+        Parameters
+        ----------
+        stats: list of dicts giving single-ifo quantities, ordered as
+            self.ifos
+        shift: numpy array of float, size of the time shift vector for each
+            coinc to be ranked
+        to_shift: list of int, multiple of the time shift to apply ordered
+            as self.ifos
+
+        Returns
+        -------
+        value: log of coinc signal rate density for the given single-ifo
+            triggers and time shifts
+        """
+        logr_s = ExpFitFgBgNormStatistic.logsignalrate(self, stats,
+                                                       shift, to_shift)
+        signal_kde = self.kde_by_tid["signal"]
+        template_kde = self.kde_by_tid["template"]
+        logr_s += [numpy.log(i/j) for i, j in zip(signal_kde, template_kde)]
```

Comment:
This looks like a very strange way of doing a calculation over an array. But I think to get the correct values of the KDEs, you need to know which template or templates is being analyzed. How does this function know which template the triggers in 'stats' come from? You need to get the information about which template we are looking at and use that to index the signal and template kde arrays.